lib/iris/analysis/stats.py

```diff
@@ -79,90 +82,93 @@ def pearsonr(
-        # The operation is non-lazy.
-        # For safety we also discard any cell-measures and ancillary-variables, to
-        # avoid cube arithmetic possibly objecting to them, or inadvertently retaining
-        # them in the result where they might be inappropriate.
-        ones_cube = cube.copy()
-        ones_cube.data = np.ones_like(cube.data)
-        ones_cube.rename("unknown")
-        ones_cube.units = 1
-        for cm in ones_cube.cell_measures():
-            ones_cube.remove_cell_measure(cm)
-        for av in ones_cube.ancillary_variables():
-            ones_cube.remove_ancillary_variable(av)
-        return ones_cube
+        dim_coords_1 = {coord.name() for coord in cube_1.dim_coords}
+        dim_coords_2 = {coord.name() for coord in cube_2.dim_coords}
+        corr_coords = list(dim_coords_1.intersection(dim_coords_2))
+
+    # Interpret coords as array dimensions.
+    corr_dims = set()
+    if isinstance(corr_coords, str):
+        corr_coords = [corr_coords]
+    for coord in corr_coords:
+        corr_dims.update(cube_1.coord_dims(coord))
+
+    corr_dims = tuple(corr_dims)
 
     # Match up data masks if required.
     if common_mask:
-        # Create a cube of 1's with a common mask.
-        if ma.is_masked(cube_2.data):
-            mask_cube = _ones_like(cube_2)
-        else:
-            mask_cube = 1.0
-        if ma.is_masked(cube_1.data):
-            # Take a slice to avoid unnecessary broadcasting of cube_2.
-            slice_coords = [
-                dim_coords_1[i]
-                for i in range(cube_1.ndim)
-                if dim_coords_1[i] not in common_dim_coords
-                and np.array_equal(
-                    cube_1.data.mask.any(axis=i), cube_1.data.mask.all(axis=i)
-                )
-            ]
-            cube_1_slice = next(cube_1.slices_over(slice_coords))
-            mask_cube = _ones_like(cube_1_slice) * mask_cube
-        # Apply common mask to data.
-        if isinstance(mask_cube, iris.cube.Cube):
-            cube_1 = cube_1 * mask_cube
-            cube_2 = mask_cube * cube_2
-            dim_coords_2 = [coord.name() for coord in cube_2.dim_coords]
-
-    # Broadcast weights to shape of cubes if necessary.
-    if weights is None or cube_1.shape == smaller_shape:
-        weights_1 = weights
-        weights_2 = weights
+        mask_1 = al.ma.getmaskarray(array_1)
+        if al is np:
+            # Reduce all invariant dimensions of mask_1 to length 1.  This avoids
+            # unnecessary broadcasting of array_2.
+            index = tuple(
+                slice(0, 1)
+                if np.array_equal(mask_1.any(axis=dim), mask_1.all(axis=dim))
+                else slice(None)
+                for dim in range(mask_1.ndim)
+            )
+            mask_1 = mask_1[index]
+
+        array_2 = _mask_array(array_2, mask_1)
+        array_1 = _mask_array(array_1, al.ma.getmaskarray(array_2))
```

Comment:
I would value @bjlittle's opinion on that: my understanding of `Resolve` is that it's currently all about metadata, so asking it to also modify data values (in the form of the masks) would be an expansion of scope.  Note this sometimes also involves changing the array type (numpy array vs masked array vs dask array).

Reply:
Actually I think this could be a bit gnarly: this procedure can increase the dimensionality of the arrays.  That's pretty straightforward here because they are _just_ arrays.  Since `Resolve` returns cubes, we would have to sort out how that increased dimensionality affects allocation of coordinates.  I'm not sure I want to go that far down the `Resolve` rabbit hole... 🐇

Reply:
Conversation offline:

We confirmed that `Resolve` DOES increase the dimensionality of `lhs_` or `rhs_cube_resolved`.

However it is still true that data management is not currently in the scope of  `Resolve`. `lhs_` and `rhs_cube_resolved` in particular are not expected to be used directly in many cases, so there is no 'contract' for getting the data right in these. What is more: The mask handling of `pearsonr()` is specialised - [see the `common_mask` parameter](https://github.com/SciTools/iris/blob/e16ea932a48a581dd7b383e388efe917341bb09c/lib/iris/analysis/stats.py#L51-L55), making it even less likely that the logic in this function would be reused elsewhere.

No further action